executor/src/witgen/data_structures/finalizable_data.rs

```diff
@@ -58,84 +129,185 @@ impl<T: FieldElement> FinalizableData<T> {
         }
     }
 
     pub fn last(&self) -> Option<&Row<T>> {
-        match self.data.last() {
-            Some(Entry::InProgress(row)) => Some(row),
-            Some(Entry::Finalized(_)) => panic!("Last row already finalized."),
-            None => None,
+        if !self.post_finalized_data.is_empty() {
+            self.post_finalized_data.last()
+        } else if !self.finalized_data.is_empty() {
+            panic!("Row already finalized");
+        } else {
+            self.pre_finalized_data.last()
         }
     }
 
     pub fn mutable_row_pair(&mut self, i: usize) -> (&mut Row<T>, &mut Row<T>) {
-        let (before, after) = self.data.split_at_mut(i + 1);
+        let (before, after) = if i + 1 < self.pre_finalized_data.len() {
+            self.pre_finalized_data.split_at_mut(i + 1)
+        } else if i < self.pre_finalized_data.len() + self.finalized_data.len() {
+            panic!("Row {i} or {} (or both) already finalized.", i + 1);
+        } else {
+            self.post_finalized_data
+                .split_at_mut(i - self.pre_finalized_data.len() - self.finalized_data.len() + 1)
+        };
         let current = before.last_mut().unwrap();
         let next = after.first_mut().unwrap();
-        match (current, next) {
-            (Entry::InProgress(current), Entry::InProgress(next)) => (current, next),
-            _ => panic!("Row {} or {} (or both) already finalized.", i, i + 1),
-        }
+        (current, next)
     }
 
-    pub fn finalize(&mut self, i: usize) -> bool {
-        if let Entry::InProgress(row) = &mut self.data[i] {
-            self.data[i] = Entry::Finalized(row.finalize(&self.column_ids));
-            true
-        } else {
-            false
+    pub fn finalize_range(&mut self, range: std::ops::Range<usize>) -> usize {
+        if range.is_empty() {
+            return 0;
         }
-    }
-
-    pub fn finalize_range(&mut self, range: impl Iterator<Item = usize>) {
-        for i in range {
-            self.finalize(i);
+        if self.finalized_data.is_empty() {
+            if !self.post_finalized_data.is_empty() {
+                self.pre_finalized_data
+                    .extend(std::mem::take(&mut self.post_finalized_data));
+            }
+            let start = std::cmp::min(range.start, self.pre_finalized_data.len());
+            let end = std::cmp::min(range.end, self.pre_finalized_data.len());
+            self.pre_finalized_data
+                .drain(start..)
+                .enumerate()
+                .for_each(|(i, row)| {
+                    if start + i < end {
+                        self.finalized_data.push(row)
+                    } else {
+                        self.post_finalized_data.push(row)
+                    }
+                });
+            end - start
+        } else {
+            // If we are asked to finalize the pre-finalized data, we just don't do it.
```

Comment:
why is that? this differs from how it behaved before no?

Reply:
Yes it differs, but just not finalizing data we were asked to is not a correctness issue, it could just be a performance (memory requirement) issue.

Reply:
But note that the memory requirements of CompactData is vastly better than before: We used to store all columns in the whole system for every machine (even if they are not used in the machine), plus we stored a full vector "head" in each row.